src/vw/src/main/scala/VowpalWabbitBase.scala

```diff
@@ -0,0 +1,410 @@
+  def getEnableCacheFile: Boolean = $(enableCacheFile)
+  def setEnableCacheFile(value: Boolean): this.type = set(enableCacheFile, value)
+
+  // abstract methods that implementors need to provide (mixed in through Classifier,...)
+  def getLabelCol: String
+  def getFeaturesCol: String
+
+  implicit class ParamStringBuilder(sb: StringBuilder) {
+    def appendParam[T](param: Param[T], option: String): StringBuilder = {
+      if (!get(param).isEmpty) {
+
+        param match {
+          case _: StringArrayParam => {
+            for (q <- get(param).get)
+              sb.append(' ').append(option).append(' ').append(q)
+          }
+          case _ => sb.append(' ').append(option).append(' ').append(get(param).get)
+        }
+      }
+
+      sb
+    }
+  }
+
+  val hashSeed = new IntParam(this, "hashSeed", "Seed used for hashing")
+  setDefault(hashSeed -> 0)
+
+  def getHashSeed: Int = $(hashSeed)
+  def setHashSeed(value: Int): this.type = set(hashSeed, value)
+
+  val numBits = new IntParam(this, "numBits", "Number of bits used")
+  setDefault(numBits -> 18)
+
+  def getNumBits: Int = $(numBits)
+  def setNumBits(value: Int): this.type = set(numBits, value)
+
+  private def createLabelSetter(df: DataFrame) = {
+    val labelColIdx = df.schema.fieldIndex(getLabelCol)
+
+    if (get(weightCol).isDefined) {
+      val weightColIdx = df.schema.fieldIndex(getWeightCol)
+      (row: Row, ex: VowpalWabbitExample) =>
+        ex.setLabel(row.getDouble(weightColIdx).toFloat, row.getDouble(labelColIdx).toFloat)
+    }
+    else
+      (row: Row, ex: VowpalWabbitExample) => ex.setLabel(row.getDouble(labelColIdx).toFloat)
+  }
+
+  /**
+    * Generate namespace info (hash, feature group, field index) for supplied columns.
+    * @param schema data frame schema to lookup column indices.
+    * @return
+    */
+  private def generateNamespaceInfos(schema: StructType): Array[NamespaceInfo] =
+    (Seq(getFeaturesCol) ++ getAdditionalFeatures)
+      .map(col => new NamespaceInfo(VowpalWabbitMurmur.hash(col, getHashSeed), col.charAt(0), schema.fieldIndex(col)))
+      .toArray
+
+  /**
+    * If multiple passes are requested and cache file is not enabled, cache the data in-memory.
+    * @param inputRows input data.
+    * @return iterator over the potentially cached data.
+    */
+  private def createDataIterator(inputRows: Iterator[Row], numPassesInSpark: Int) = {
+    if (numPassesInSpark == 1)
+    // avoid caching inputRows
+      () => inputRows
+    else {
+      // cache all row references to be able to quickly get the iterator again
+      val inputRowsArr = inputRows.toArray
+      () => inputRowsArr.iterator
+    }
+  }
+
+  private def buildCommandLineArguments(vwArgs: String, contextArgs: () => String = () => "") = {
+    val args = new StringBuilder
+    args.append(vwArgs)
+      .append(" ").append(contextArgs())
+
+    // have to pass to get multi-pass to work
+    if (args.indexOf("--no_stdin") == -1)
```

Comment:
Consider keeping the args + arg logic in the types until the last minute then move them all to the training string